### PR TITLE
Respect newlines in data declarations in more cases

### LINKED ISSUE #947: Line breaks ignored in data type declarations.
**Describe the bug**
Ormolu sometimes eats line breaks within data type declarations.

**To Reproduce**
 Run ormolu on code similar to the following, with a line break before the `=`:
 
 ```hs
 data SomeReasonablyDescriptiveRecordType someParam1 someParam2 someParam3
  = SomeReasonablyDescriptiveRecordType
    { foo :: someParam1,
      bar :: someParam2,
      baz :: someParam3
    }
```
or with the line break after the `=`:
 ```hs
 data SomeReasonablyDescriptiveRecordType someParam1 someParam2 someParam3 =
  SomeReasonablyDescriptiveRecordType
    { foo :: someParam1,
      bar :: someParam2,
      baz :: someParam3
    }
``` 

**Expected behaviour**
The line break is preserved, and the data constructor name is indented one level to the right.

**Actual behaviour**
Ormolu eats the line break, and prints the data constructor name on the same line, resulting in a rather long line:
```hs
data SomeReasonablyDescriptiveRecordType someParam1 someParam2 someParam3 = SomeReasonablyDescriptiveRecordType
{ foo :: someParam1,
  bar :: someParam2,
  baz :: someParam3
}
```

**Environment**
 - OS name + version: `Linux 5.4.0-135-generic #152-Ubuntu SMP Wed Nov 23 20:19:22 UTC 2022 x86_64 GNU/Linux`
 - Version of the code: `ormolu 0.5.1.0 UNKNOWN UNKNOWN using ghc-lib-parser 9.4.3.20221104`

**Additional context**
- We have a rather large code base (> 200,000 lines) that's formatted by hand, and we're interested in adopting an auto-formatter.
- Our team coding standard encourages an 80-column limit.

We realise that ormolu does not provide support for a column limit, but we're wondering if we can get by with manually inserting line breaks where necessary.

Many thanks!
Jonathan

### LINKED ISSUE #1077: Format existential data types better
I think this should be the correct formatting:
```hs
data IndexWithInfo schema =
  forall x.
  IndexWithInfo
    { checkedIndex :: Index schema x
    , checkedIndexName :: U.Variable
    , checkedIndexType :: Type x
    }
```
Currently, Ormolu reformats this to
```hs
data IndexWithInfo schema = forall x.
  IndexWithInfo
  { checkedIndex :: Index schema x
  , checkedIndexName :: U.Variable
  , checkedIndexType :: Type x
  }
```

### PR DESCRIPTION
**Pull request by mrkkrp**
Close #1077, close #947.

### CODE REVIEW

File: data/examples/declaration/data/field-layout/record-1-out.hs [hunk 1]
```diff
@@ -0,0 +1,11 @@
...
+data Foo
+  = Foo
+  { -- | Something
+    foo :: Foo Int Int,
+    -- | Something else
+    bar ::
+      Bar
+        Char
+        Char
+  }
```
Comment (amesgen) [inline data/examples/declaration/data/field-layout/record-1-out.hs:11]: Hmm, the `{}` brackets enclosing the fields are at the same indentation level as the `=`, but if you add another constructor in the end, then they are indented by two more steps:
```haskell
data Foo
  = Foo
      { -- | Something
        foo :: Foo Int Int,
        -- | Something else
        bar ::
          Bar
            Char
            Char
      }
  | Bar
```
This feels a bit inconsistent, but also not like a huge deal. WDYT?

File: src/Ormolu/Printer/Meat/Declaration/Data.hs [hunk 2]
```diff
@@ -50,74 +50,78 @@ p_dataDecl style name tyVars getTyVarLoc p_tyVar fixity HsDataDefn {..} = do
...
         NewTypeCon a -> [a]
         DataTypeCons _ as -> as
       gadt = isJust dd_kindSig || any (isGadt . unLoc) dd_cons'
-  unless (null dd_cons') $
-    if gadt
-      then inci $ do
-        switchLayout declHeaderSpans $ do
+  case dd_cons' of
+    [] -> pure ()
+    (first_dd_cons : _) ->
```
Comment (amesgen) [inline src/Ormolu/Printer/Meat/Declaration/Data.hs:90]:
```suggestion
    first_dd_cons : _ ->
```

**Review by amesgen — COMMENTED** [review]
This looks great, just one question about non-compact single constructor ADTs.

File: data/examples/declaration/data/field-layout/record-1-out.hs (hunk 1 above)
Comment (mrkkrp) [reply data/examples/declaration/data/field-layout/record-1-out.hs:11]: Yeah, that was a decision that I made a while ago. A single constructor records are the default so it made sense to save a little bit of horizontal space by not aligning them. This PR simply keeps the status quo.

**Review by amesgen — APPROVED** [review]

**Pull request merged**